Implemented guild experiments

BanTheNons opened the pull request.

BanTheNons committed:
Squashed commit of the following:

commit e587b8f81f366b175a042ef1eb96e44bc8807570
Author: BuildTools <59115290+BanTheNons@users.noreply.github.com>
Date:   Tue Sep 21 19:19:57 2021 +0300

    Create guild_experiments.example.json

commit 9cb03c85618d62e3501146e96539e2c0c79416a3
Author: BuildTools <59115290+BanTheNons@users.noreply.github.com>
Date:   Tue Sep 21 22:32:07 2021 +0300

    Update Identify.ts

commit 3b0e0bb116841c460f41b7429a59a577f61e3863
Author: BanTheNons <59115290+BanTheNons@users.noreply.github.com>
Date:   Tue Sep 21 16:56:14 2021 +0300

    implemented guild experiments

commit 489af5182d8c69f46b468f010480c88565fbb42f
Author: BuildTools <59115290+BanTheNons@users.noreply.github.com>
Date:   Tue Sep 21 22:26:53 2021 +0300

    Update .gitignore

BanTheNons committed:
Update guild_experiments.json

BanTheNons committed:
compile json files

DiegoMagdaleno commented:
Maybe we should wait for FOSScord to have proper support to handle some of this experiments?

DiegoMagdaleno commented on api/src/routes/experiments.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "experiments": {"fingerprint": "889800254663311382.rgUouuEkLbQi0SGiEMClnfgNZic", "assignments": [[4047587481, 0, 0, -1, 0], [1865079242, 0, 1, -1, 0], [3816091942, 3, 0, -1, 0], [4130837190, 0, 10, -1, 0], [1861568052, 0, 1, -1, 0], [1578940118, 1, 1, -1, 0], [1571676964, 0, 1, -1, 2], [1658164312, 2, 1, -1, 0], [98883956, 1, 1, -1, 0], [3114091169, 0, 1, -1, 0], [2570684145, 4, 1, -1, 2], [4 ...
```
Is this the only way to do it?.

It doesn't look very readable.

DiegoMagdaleno commented on api/src/routes/experiments.ts:
```diff
@@ -1,11 +1,13 @@
 router.get("/", route({}), (req: Request, res: Response) => {
-	// TODO:
-	res.send({ fingerprint: "", assignments: [] });
+	console.log(req.params)
```
Please don't "console.log()" as it can expose some information

BanTheNons replied:
Forgot to remove that!

DiegoMagdaleno commented on api/src/routes/experiments.ts:
```diff
@@ -1,11 +1,13 @@
-	// TODO:
-	res.send({ fingerprint: "", assignments: [] });
+	console.log(req.params)
+	if (req.url.includes("with_guild_experiments=true")) return res.json(with_guild_experiments);
```
Could this be a ternary operator?, like this:

`return (req.url.includes("with_guild_experiments=true") ? res.json(with_guild_experiments) : res.json(experiments))`

I don't know if that would work, so it would be appreciated if you could try it

Flam3rboy replied:
yes ternary is possible, but the current notation is cleaner imo (tho the else should be removed, as it's unnecessary).

DiegoMagdaleno replied:
I think Ternary is cleaner when reading ifs that are one line tho.

Flam3rboy replied:
Thats a matter of opinion, tho the code style should be consistent and therefore we should use the if with return

DiegoMagdaleno replied:
Okay, ill update the review

Flam3rboy commented on api/src/routes/experiments.ts:
```diff
@@ -1,11 +1,13 @@
-	res.send({ fingerprint: "", assignments: [] });
+	console.log(req.params)
+	if (req.url.includes("with_guild_experiments=true")) return res.json(with_guild_experiments);
+	else return res.json(experiments);
```
remove the else as its unecssary

DiegoMagdaleno reviewed (commented):
Just some issues with the code style, once those are fixed, it is good to go.

erkinalp commented:
> Maybe we should wait for FOSScord to have proper support to handle some of this experiments?

Something like `/policies/features/guilds/:id`?

BanTheNons commented:
> Something like `/policies/features/guilds/:id`, so that the server operator could configure features per guild?

Guild IDs can be added as overrides in the guild_experiments.json file, if you could only enable experiments per guild you wouldn't be able to set rollouts for it.

BanTheNons committed:
Update experiments.ts

BanTheNons committed:
Delete guild_experiments.json

You can just rename guild_experments.example.json and you should be good to go.

samuelscheit committed:
Merge branch 'master' into pr/BanTheNons/389

Flam3rboy reviewed (changes requested):
We shouldn't save the experiment overrides in a file, as it would create inconsistency between the gateway and api and if you would deploy multiple instances.
Instead we should save them in the database as a simple JSON string (for now).

erkinalp reviewed (changes requested):
Rebase against updated dependencies

TheArcaneBrony commented:
was accidentally closed
